File: innerlibs1/DOT Language.vb
Namespace DOTLanguage

    Enum GraphType
        ''' <summary>
        ''' Gráficos não orientados
        ''' </summary>
        Graph
        ''' <summary>
        ''' Gráficos orientados
        ''' </summary>
        Digraph
    End Enum

    ''' <summary>
    ''' Wrapper para criaçao de gráficos em DOT Language
    ''' </summary>
    Public Class Graph
        Inherits List(Of DotObject)

        Property Clusters As New List(Of Cluster)

        ''' <summary>
        ''' Tipo do Grafico (graph, digraph)
        ''' </summary>
        ''' <returns></returns>

        Property GraphType As String = "graph"

        Property Strict As Boolean = False

        ''' <summary>
        ''' Nome do Gráfico
        ''' </summary>
        ''' <returns></returns>
        Property ID As String = ""

        ''' <summary>
        ''' Escreve a DOT string correspondente a este gráfico
        ''' </summary>
        ''' <returns></returns>
        Public Overrides Function ToString() As String
            Dim s = Me.Select(Function(n) n.ToString & Environment.NewLine).ToArray.Join("")
            s = s.Split(Environment.NewLine).Distinct.Join(Environment.NewLine) & Environment.NewLine
            If GraphType.ToLower.Equals("graph") Then
                s = s.Replace("->", "--").Replace("<-", "--")
            End If
            Return GraphType & " " & ID.ToSlugCase(True) & " " & s.Quote("{")
        End Function

    End Class

    Public Class Cluster
        Inherits DotObject

        Public Overrides Property ID As String
            Get
                Throw New NotImplementedException()
            End Get
            Set(value As String)
                Throw New NotImplementedException()
            End Set
        End Property

    End Class

    Public MustInherit Class DotObject

        Public MustOverride Property ID As String

        Public ReadOnly Property Attributes As New DotAttributeCollection()

    End Class

    Public Class DotAttributeCollection
        Inherits Dictionary(Of String, Object)

        Public Overrides Function ToString() As String
            Dim dotstring = ""
            For Each prop In Me
                Dim val = prop.Value.ToString.QuoteIf(prop.Value.ToString.Contains(" ") Or prop.Value.ToString.IsBlank Or prop.Value.ToString.IsURL)
                If val.IsIn({"True", "False"}) Then val = val.ToLower
                If val.IsNumber Then val = val.ChangeType(Of Decimal).ToString("00,00")
                dotstring  &=  prop.Key & "=" & val & " "
            Next
            Return dotstring.Quote("[") & ";" & Environment.NewLine
        End Function

    End Class

    ''' <summary>
    ''' Representa um nó de um grafico em DOT Language
    ''' </summary>
    Public Class DotNode
        Inherits DotObject

        ''' <summary>
        ''' Cria um novo nó
        ''' </summary>
        ''' <param name="ID"></param>
        Sub New(ID As String)
            Me.ID = ID
        End Sub

        ''' <summary>
        ''' ID deste nó
        ''' </summary>
        ''' <returns></returns>
        Public Overrides Property ID As String
            Get
                Return _id
            End Get
            Set(value As String)
                _id = value.ToSlugCase(True)
            End Set
        End Property

        Private _id As String

        ''' <summary>
        ''' Escreve a DOT string deste nó e seus respectivos nós filhos
        ''' </summary>
        ''' <returns></returns>
        Public Overrides Function ToString() As String
            Return ID & Me.Attributes.ToString & Environment.NewLine
        End Function

    End Class

    ''' <summary>
    ''' Representa uma ligação entre nós de um grafico em DOT Language
    ''' </summary>
    Public Class DotEdge
        Inherits DotObject

        ''' <summary>
        ''' Cria uma nova ligaçao
        ''' </summary>
        ''' <param name="Oriented">Relação orientada</param>
        Sub New(ParentNode As DotNode, ChildNode As DotNode, Optional Oriented As Boolean = True)
            Me.ParentNode = ParentNode
            Me.ChildNode = ChildNode
            Me.Oriented = Oriented
        End Sub

        ''' <summary>
        ''' Indica se esta ligação é orientada ou não
        ''' </summary>
        ''' <returns></returns>
        Property Oriented As Boolean = True

        Property ParentNode As DotNode

        Property ChildNode As DotNode

        Public Overrides Property ID As String
            Get
                Return ParentNode.ID.ToSlugCase(True) & If(Me.Oriented, " -> ", " -- ") & ChildNode.ID.ToSlugCase(True)
            End Get
            Set(value As String)
                Debug.Write("Cannot change ID of a relation")
            End Set
        End Property

        ''' <summary>
        ''' Escreve a DOT String desta ligaçao
        ''' </summary>
        ''' <returns></returns>
        Public Overrides Function ToString() As String
            Dim dotstring = ""
            If Me.Attributes.Any() Then
                dotstring = Me.ID & " " & Me.Attributes.ToString & Environment.NewLine
            End If
            Return dotstring
        End Function

    End Class

End Namespace

File: InnerLibs/DOT Language.cs
using InnerLibs.LINQ;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace InnerLibs.DOTLanguage
{
    public class Cluster : DotObject
    {
        #region Public Properties

        public override string ID
        {
            get;
            set;
        }

        #endregion Public Properties
    }

    public class DotAttributeCollection : Dictionary<string, object>
    {
        #region Public Methods

        public override string ToString()
        {
            string dotstring = InnerLibs.Text.Empty;
            foreach (var prop in this)
            {
                string val = prop.Value.ToString().QuoteIf(prop.Value.ToString().Contains(" ") | prop.Value.ToString().IsBlank() | prop.Value.ToString().IsURL());
                if (Misc.IsIn(val, new[] { "True", "False" }))
                {
                    val = val.ToLowerInvariant();
                }

                if (val.IsNumber())
                {
                    val = val.ChangeType<decimal>().ToString("00,00");
                }

                dotstring += prop.Key + "=" + val + " ";
            }

            return dotstring.Quote('[') + ";" + Environment.NewLine;
        }

        #endregion Public Methods
    }

    /// <summary>
    /// Representa uma ligação entre nós de um grafico em DOT Language
    /// </summary>
    public class DotEdge : DotObject
    {
        #region Public Constructors

        /// <summary>
        /// Cria uma nova ligação
        /// </summary>
        /// <param name="Oriented">Relação orientada</param>
        public DotEdge(DotNode ParentNode, DotNode ChildNode, bool Oriented = true)
        {
            this.ParentNode = ParentNode;
            this.ChildNode = ChildNode;
            this.Oriented = Oriented;
        }

        #endregion Public Constructors

        #region Public Properties

        public DotNode ChildNode { get; set; }

        public override string ID
        {
            get => ParentNode.ID.ToSlugCase(true) + (Oriented ? " -> " : " -- ") + ChildNode.ID.ToSlugCase(true);

            set => Misc.WriteDebug("Cannot change ID of a relation");
        }

        /// <summary>
        /// Indica se esta ligação é orientada ou não
        /// </summary>
        /// <returns></returns>
        public bool Oriented { get; set; } = true;

        public DotNode ParentNode { get; set; }

        #endregion Public Properties

        #region Public Methods

        /// <summary>
        /// Escreve a DOT String desta ligação
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            string dotstring = InnerLibs.Text.Empty;
            if (Attributes.Any())
            {
                dotstring = ID + " " + Attributes.ToString() + Environment.NewLine;
            }

            return dotstring;
        }

        #endregion Public Methods
    }

    /// <summary>
    /// Representa um nó de um grafico em DOT Language
    /// </summary>
    public class DotNode : DotObject
    {
        #region Private Fields

        private string _id;

        #endregion Private Fields

        #region Public Constructors

        /// <summary>
        /// Cria um novo nó
        /// </summary>
        /// <param name="ID"></param>
        public DotNode(string ID)
        {
            this.ID = ID;
        }

        #endregion Public Constructors

        #region Public Properties

        /// <summary>
        /// ID deste nó
        /// </summary>
        /// <returns></returns>
        public override string ID
        {
            get => _id;

            set => _id = value.ToSlugCase(true);
        }

        #endregion Public Properties

        #region Public Methods

        /// <summary>
        /// Escreve a DOT string deste nó e seus respectivos nós filhos
        /// </summary>
        /// <returns></returns>
        public override string ToString() => ID + Attributes.ToString() + Environment.NewLine;

        #endregion Public Methods
    }

    public abstract class DotObject
    {
        #region Public Properties

        public DotAttributeCollection Attributes { get; private set; } = new DotAttributeCollection();
        public abstract string ID { get; set; }

        #endregion Public Properties
    }

    /// <summary>
    /// Wrapper para criação de gráficos em DOT Language
    /// </summary>
    public class Graph : List<DotObject>
    {
        #region Public Properties

        public List<Cluster> Clusters { get; set; } = new List<Cluster>();

        /// <summary>
        /// Tipo do Grafico (graph, digraph)
        /// </summary>
        /// <returns></returns>
        public GraphType GraphType { get; set; } = GraphType.Graph;

        /// <summary>
        /// Nome do Gráfico
        /// </summary>
        /// <returns></returns>
        public string ID { get; set; } = InnerLibs.Text.Empty;

        public bool Strict { get; set; } = false;

        #endregion Public Properties

        #region Public Methods

        /// <summary>
        /// Escreve a DOT string correspondente a este gráfico
        /// </summary>
        /// <returns></returns>
        public override string ToString()
        {
            var gg = $"{GraphType}".ToLowerInvariant();
            string s = this.Select(n => n.ToString() + Environment.NewLine).ToArray().SelectJoinString();
            s = s.Split(Environment.NewLine).Distinct().SelectJoinString(Environment.NewLine) + Environment.NewLine;
            if (gg.Equals("graph"))
            {
                s = s.Replace("->", "--").Replace("<-", "--");
            }

            return gg + " " + ID.ToSlugCase(true) + " " + s.Quote('{');
        }

        #endregion Public Methods
    }

    public enum GraphType
    {
        /// <summary>
        /// Gráficos não orientados
        /// </summary>
        Graph,

        /// <summary>
        /// Gráficos orientados
        /// </summary>
        Digraph
    }
}